be/src/exec/pipeline/hash_partition_context.cpp

```diff
@@ -0,0 +1,101 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/pipeline/hash_partition_context.h"
+
+#include "exprs/expr.h"
+
+namespace starrocks::pipeline {
+
+Status HashPartitionContext::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(Expr::create_expr_trees(state->obj_pool(), _t_partition_exprs, &_partition_exprs, state));
+    RETURN_IF_ERROR(Expr::prepare(_partition_exprs, state));
+    RETURN_IF_ERROR(Expr::open(_partition_exprs, state));
+    for (auto& expr : _partition_exprs) {
+        auto& type_desc = expr->root()->type();
+        if (!type_desc.support_groupby()) {
+            return Status::NotSupported(fmt::format("partition by type {} is not supported", type_desc.debug_string()));
+        }
+    }
+    auto partition_size = _t_partition_exprs.size();
+    _partition_types.resize(partition_size);
+    for (auto i = 0; i < partition_size; ++i) {
+        TExprNode expr = _t_partition_exprs[i].nodes[0];
+        _partition_types[i].result_type = TypeDescriptor::from_thrift(expr.type);
+        _partition_types[i].is_nullable = expr.is_nullable;
+        _has_nullable_key = _has_nullable_key || _partition_types[i].is_nullable;
+    }
+
+    _chunks_partitioner = std::make_unique<ChunksPartitioner>(_has_nullable_key, _partition_exprs, _partition_types);
+    return _chunks_partitioner->prepare(state);
+}
+
+Status HashPartitionContext::push_one_chunk_to_partitioner(RuntimeState* state, const ChunkPtr& chunk) {
+    return _chunks_partitioner->offer<false>(chunk, nullptr, nullptr);
+}
+
+void HashPartitionContext::sink_complete() {
+    _is_sink_complete = true;
+}
+
+bool HashPartitionContext::has_output() {
+    return _is_sink_complete && (!_chunks_partitioner->is_hash_map_eos() || _chunk != nullptr);
+}
+
+bool HashPartitionContext::is_finished() {
+    if (!_is_sink_complete) {
+        return false;
+    }
+    return !has_output();
+}
+
+StatusOr<ChunkPtr> HashPartitionContext::pull_one_chunk(RuntimeState* state) {
+    ChunkPtr output = nullptr;
+    if (!_chunks_partitioner->is_hash_map_eos()) {
+        _chunks_partitioner->consume_from_hash_map([&](int32_t partition_idx, ChunkPtr& chunk) {
+            if (_chunk == nullptr) {
+                if (chunk->num_rows() < state->chunk_size()) {
+                    _chunk = std::move(chunk);
+                    return true;
+                } else {
+                    output = std::move(chunk);
+                    return false;
+                }
+            } else if (_chunk->num_rows() + chunk->num_rows() <= state->chunk_size()) {
+                _chunk->append(*chunk);
+                chunk = nullptr;
+                return true;
+            } else {
+                output = std::move(_chunk);
+                _chunk = std::move(chunk);
+                return false;
+            }
```

Comment:
no needs to write duplicate code for accummulate small chunks into large one, suggest to use ChunkAccummulator.

Reply:
Good point, I'll update.